resolver/map.go

```diff
@@ -136,3 +136,134 @@ func (a *AddressMap) Values() []any {
+	for _, addr := range endpoint.Addresses {
+		if count, ok := en[addr.Addr]; ok {
+			en[addr.Addr] = count + 1
+		}
+		en[addr.Addr] = 1
+	}
+	return endpointNode{
+		addrs: en,
+	}
+}
+
+// EndpointMap is a map of endpoints to arbitrary values keyed on only the
+// unordered set of addresses string within an endpoint. The reason this uses a
+// slice instead of a map is because slices or maps cannot be used as map keys.
+// Thus, for efficiency store as a map and compare to others so each node
+// comparison is o(n), where n is the number of unique addresses in the endpoint
+// rather than having to sort a list on each comparison which is n(log(n))
+// (cannot use slices for the latter case or maps for the former case as map
+// keys, so needs to be a list).
+type EndpointMap struct {
+	endpoints endpointNodeList // Cannot use maps as a map key. Thus, keep a list, and o(n) search. Not on RPC fast path so efficiency not as important.
+}
+
+// NewEndpointMap returns a new EndpointMap.
+func NewEndpointMap() *EndpointMap {
+	return &EndpointMap{}
+}
+
+// Get returns the value for the address in the map, if present.
+func (em *EndpointMap) Get(e Endpoint) (value any, ok bool) {
+	en := toEndpointNode(e)
+	if i := em.endpoints.find(en); i != -1 {
+		return em.endpoints[i].val, true
+	}
+	return nil, false
+}
+
+// Set updates or adds the value to the address in the map.
+func (em *EndpointMap) Set(e Endpoint, value any) {
+	en := toEndpointNode(e)
+	if i := em.endpoints.find(en); i != -1 {
+		em.endpoints[i].val = value
+		return
+	}
+	en.val = value
+	em.endpoints = append(em.endpoints, en)
+}
+
+// Len returns the number of entries in the map.
+func (em *EndpointMap) Len() int {
+	return len(em.endpoints)
+}
+
+// Keys returns a slice of all current map keys, as endpoints specifying the
+// addresses present in the endpoint keys, in which uniqueness is determined by
+// the unordered set of addresses. Thus, endpoint information returned is not
+// the full endpoint data but can be used for EndpointMap accesses.
+func (em *EndpointMap) Keys() []Endpoint { // TODO: Persist the whole endpoint data to return in nodes? No use case now, but one could come up in future.
+	ret := make([]Endpoint, 0, len(em.endpoints))
+	for _, en := range em.endpoints {
+		var endpoint Endpoint
+		for addr, count := range en.addrs {
+			for i := 0; i < count; i++ {
+				endpoint.Addresses = append(endpoint.Addresses, Address{
+					Addr: addr,
+				})
+			}
+		}
+		ret = append(ret, endpoint)
+	}
+	return ret
+}
+
+// Values returns a slice of all current map values.
+func (em *EndpointMap) Values() []any {
+	ret := make([]any, 0, len(em.endpoints))
+	for _, en := range em.endpoints {
+		ret = append(ret, en.val)
+	}
+	return ret
+}
+
+// find returns the index of endpoint in the endpointNodeList slice, or -1 if
+// not present. The comparisons are done on the unordered set of the addresses
+// within endpoint.
+func (l endpointNodeList) find(e endpointNode) int {
```

Comment:
Suggestion: remove this type as it obfuscates that it's a slice.  Implement this method on `EndpointMap` instead.

Reply:
Hmmmm good idea. Switched.